packages/jaeger-ui/src/components/TracePage/TracePageHeader/SpanGraph/CanvasSpanGraph.tsx

```diff
@@ -21,7 +21,14 @@ import { TNil } from '../../../../types';
 import './CanvasSpanGraph.css';
 
 type CanvasSpanGraphProps = {
-  items: { valueWidth: number; valueOffset: number; serviceName: string }[];
+  // TODO: Unclear why this is not the `SpanItem` type
+  items: {
+    valueWidth: number;
+    valueOffset: number;
+    serviceName: string;
+    spanGroup: string;
+    rgbColor: [number, number, number];
```

Comment:
I am not sure I follow this:
  * why do we need both spanGroup and color?
  * shouldn't color be always calculated rather than captured in the item?

Reply:
`spanGroup` is the key in `fillCache` that `renderIntoCanvas` uses, `rgbColor` is the color to use.

This could go either way, I think. The previous code didn't pass in the colour, instead, it passed in a function to get the colour (`getColor`, removed here from line 53).

To me, that felt a bit weird -- the caller of this code knows both the function to call, and the value of the parameter it will be called with. So it made sense to move that responsibility in to `getItem` in `SpanGraph/index.tsx`.

I actually think the best way to do this is to make the color a property on the `Span` type, populate it when a span is created, and have `getItem` fetch the color the same way it fetches the other `Span` properties.

I can move it back if you prefer though.

When I was first navigating this code I found the indirection through the `getItem` function confusing. For example, I don't understand the design tradeoff between making it a function vs. an `asSpanItem()` method on the `Span` type.

And (as noted in the comment in the code here) why `items` explicitly lists the fields it accepts, instead of just being `items: SpanItem[]`.

Or (and this would also be a larger refactoring) why `renderIntoCanvas` takes an array of `SpanItem` and not an array of `Span`. The `SpanItem` type is a strict subset of the `Span` type, so this seems like an unnecessary additional level of indirection (unless maybe it makes tests easier...? I haven't checked there yet).

Reply:
I don't know the answers to all "why"s, but some general design considerations I could offer are:
  * Color is not a property of the span, it is a function of the span attributes and a given display mode. Span contains the data, that same data could be displayed differently in different views or based on different use selections.
  * Different components / parts of the code declare "span" only as wide as they need it to be. The type declaration defines the contract that the code expects the data to satisfy. This decouples different components from depending on each other or on the shared data model.